### PR TITLE
Add CloudWatchLogsClient module

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Add `CloudWatchLogsClient` module that can retrieve logs for ECS task runs of SAP invoices processing.`CloudWatchLogsClient.get_log_messages` returns a list of strings where each string represents a logged message; the method relies on [`boto3.ECS.Client.get_log_events`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs/client/get_log_events.html). 

While addressing initial set of review comments, additional required updates were made based on how the `CloudWatchLogsClient` is used in the context of the Alma SAP Invoices Flask app.

**Note(s):**
* A new environment variable `ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP` is introduced.

### How can a reviewer manually see the effects of these changes?

1. Run `make test` and verify all tests are passing.
Note: For now, running the preliminary set of unit tests is the best way to confirm that the implemented `webapp.utils.aws.cloudwatch.CloudWatchLogsClient` module is working as expected.

### Includes new or updated dependencies?
YES - Installed **dev** packages: [`boto3-stubs["logs"]`, `pandas`].

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/IN-1015

### Developer
- [x] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [x] All related Jira tickets are linked in commit message(s)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: webapp/utils/aws/cloudwatch.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        logger.info("Retrieving CloudWatch logs for task.")
+        log_events = []
+        params = {
+            "logGroupName": self.log_group_name,
+            "logStreamName": f"{self.log_stream_name_prefix}{task_id}",
+            "startFromHead": True,
+        }
+
+        while True:
+            response = self.client.get_log_events(**params)  # type: ignore[arg-type]
```
Comment (ghukill) [inline webapp/utils/aws/cloudwatch.py:40]: What happens here if the log group/stream does not exist?  The [documentation here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs/client/get_log_events.html#) suggests that `CloudWatchLogs.Client.exceptions.ResourceNotFoundException` can get thrown if that's the case.

Would it be worth considering adding a check _before_ this `while` loop to ensure the log stream exists before attemping to retrieve lines?  Could even be a method like `log_stream_exists()`?  

Or... perhaps this work would come later when integrating into the flow of things?  I could a see a universe where we kickoff the ECS task, but the log stream hasn't _yet_ been created, but we are immediately querying for it.

Maybe don't need to solve now, just noting.

**Review by ghukill — APPROVED** [review]
Looks great!  Left a couple of optional questions and suggestions, but not blocking for approval.  

I found the test fixtures particularly illuminating, both in _how_ to mock AWS logs, and how the log group/stream structure is driven by the ECS task id.

File: webapp/utils/aws/cloudwatch.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            log_events.extend(response["events"])
+            next_token = response.get("nextForwardToken")
+            if next_token == params.get("nextToken"):
+                break
+            params["nextToken"] = next_token
+        logger.info("CloudWatch logs retrieved.")
+        return self.get_messages_from_logs(log_events)
+
+    def get_messages_from_logs(self, logs: list) -> list[str]:
+        return [event["message"] for event in logs]
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply webapp/utils/aws/cloudwatch.py:50]: I see some value in that pattern as well if only for one less variable to pass around (unless something is expected to happen that variable between `get_logs` and `get_messages_from_logs`)

**Review by ehanson8 — COMMENTED** [review]
A few minor thoughts that mostly piggyback off of Graham's suggestions

File: webapp/utils/aws/cloudwatch.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    from mypy_boto3_logs.client import CloudWatchLogsClient as CloudWatchLogsClientType
+
+from webapp.config import Config
+
+logger = logging.getLogger(__name__)
+
+
+@define
+class CloudWatchLogsClient:
+    """ECS Client for running and monitoring task runs."""
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/aws/cloudwatch.py]: T'was indeed. I updated the [class docstring](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R17-R26), which includes additional info recommended by @ghukill (see [comment](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12#discussion_r1676250737)).

[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/aws/cloudwatch.py]: Thanks! I also wanted to clarify that the `startFromHead` parameter does not indicate the "offset" in the log events 🤔 --here is the description from [boto3 documentation on `CloudWatchLogs.Client.get_log_events`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs/client/get_log_events.html):
> startFromHead (boolean) –
> If the value is true, the earliest log events are returned first. If the value is false, the latest log events are returned first. The default value is false.
>If you are using a previous nextForwardToken value as the nextToken in this operation, you must specify true for startFromHead.
TLDR; `startFromHead` determines how the log events should be sorted!

File: webapp/utils/aws/cloudwatch.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/aws/cloudwatch.py]: Thank you for the suggestions! I updated the `CloudWatchLogsClient` to include the following:
* [get_log_messages](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R39-R43): Get messages from log events.
* [get_log_events](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R45-R65): Get log events. 
* [get_log_streams](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R73-R88): Get log stream _names_ for the log group indicated by `CloudWatchLogsClient.log_group_name`.

Read [AWS User Guide | Amazon CloudWatch Logs concepts](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CloudWatchLogsConcepts.html) for more info!

File: tests/conftest.py [hunk 4]
```diff
@@ -132,12 +143,110 @@ def ecs_client_execute_run_details_success(
...
+        the "messages" column (i.e., the logged messages normally lead with a
+        timestamp formatted like so "2024-07-02 13:58:20,524").
+    """
+    # create log group and log stream
+    logs = boto3.client("logs", region_name=AWS_DEFAULT_REGION)
+    task_id = "abc001"
+    log_stream_name = f"sapinvoices/mock-sapinvoices-ecs-test/{task_id}"
+    logs.create_log_stream(
+        logGroupName=mock_cloudwatchlogs_log_group, logStreamName=log_stream_name
+    )
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply tests/conftest.py:199]: I am glad you find it helpful!

File: webapp/utils/aws/cloudwatch.py (hunk 1 above)
Comment (jonavellecuerdo) [reply webapp/utils/aws/cloudwatch.py]: I decided to:

[1. Write a method to check if a log stream exists by doing performing an "exact match" of a given task ID against the task IDs retrieved from the log stream names retrieved by `CloudWatchLogsClient.get_log_streams`](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R67-R71)

[2. Use the method from 1 in `CloudWatchLogsClient.get_log_events` and raise a custom exception `ECSTaskLogStreamDoesNotExistError` when a log stream for the task ID is not found](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R54-R65). 

**Note:** I opted to create a custom exception as it allows our exceptions to have consistent naming and write cleaner messages. In the case of the `ResourceNotFoundException`, I would've had to write the `try-except` clause with the following syntax:

```
try;
   <try this>
except self.client.exceptions.ResourceNotFoundException as exception:
   <do this>
```

Because the exception needs to be imported _from_ the `boto3` client, it feels a little confusing within the `CloudWatchLogsClient`. 

Let me know what you think!

File: webapp/utils/aws/cloudwatch.py [hunk 5]
```diff
@@ -27,7 +36,13 @@ class CloudWatchLogsClient:
     def client(self) -> "CloudWatchLogsClientType":
         return boto3.client("logs")
 
-    def get_logs(self, task_id: str) -> list[str] | None:
+    def get_log_messages(self, task_id: str) -> list:
+        messages: list = []
+        if logs := self.get_log_events(task_id):
+            messages.extend(event["message"] for event in logs)
+        return messages
```
Comment (ehanson8) [inline webapp/utils/aws/cloudwatch.py:43]: Could this `yield` or does it need to be a list?

**Review by ehanson8 — COMMENTED** [review]
Looks great but 1 question!

File: webapp/utils/aws/cloudwatch.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    from mypy_boto3_logs.client import CloudWatchLogsClient as CloudWatchLogsClientType
+
+from webapp.config import Config
+from webapp.exceptions import ECSTaskLogStreamDoesNotExistError
+
+logger = logging.getLogger(__name__)
+
+
+@define
+class CloudWatchLogsClient:
```
Comment (ghukill) [inline webapp/utils/aws/cloudwatch.py:17]: Short of getting wired up to routes and logic in the actual flask app, this class looks great!  Simple methods, makes sense.

**Review by ghukill — APPROVED** [review]
Approved!  Cloudwatch client looks great, makes sense.  

I did leave a question about `Config` properties that seem to be mirroring env vars, but not a blocker.

File: webapp/utils/aws/cloudwatch.py (hunk 5 above)
Comment (jonavellecuerdo) [reply webapp/utils/aws/cloudwatch.py:43]: It needs to be a list. You can view this [sample function call in the code under development](https://github.com/MITLibraries/alma-sapinvoices-ui/blob/frontend-dev/webapp/app.py#L47) to see how `CloudWatchLogsClient` is used by the Flask app.

File: webapp/config.py [hunk 7]
```diff
@@ -59,6 +60,10 @@ def ALMA_SAP_INVOICES_ECS_CLUSTER(self) -> str:
     def ALMA_SAP_INVOICES_ECS_TASK_DEFINITION(self) -> str:
         return self._getenv("ALMA_SAP_INVOICES_ECS_TASK_DEFINITION")
 
+    @property
+    def ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP(self) -> str:
+        return self._getenv("ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP")
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/config.py:65]: Ah, traveled down memory lane and we did have [some discussions on why these `@property` methods exist](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/1#discussion_r1664403672) to support accessing environment variables from a Flask `app.config` object. That said, I think it's worth revisiting but would like to push that body of work to a later PR if that's okay!

**Comment by jonavellecuerdo** [comment]
Just to follow up: thanks to Aaron's help, the Lambda function URL is operational: https://ajpu5ef4idkhdtss2whd4tgdyi0gncpr.lambda-url.us-east-1.on.aws/. :rocket: 

The app that launches represents the [latest code under development.](https://github.com/MITLibraries/alma-sapinvoices-ui/tree/frontend-dev)

1. If you select `Review Run` or `Final Run` you will see an output similar to the following:
   image

   Note: The "logs" shown are valid and are the log messages sent to CloudWatch when we try to execute a run of the [Alma SAP Invoices task with _invalid_ overrides](https://github.com/MITLibraries/alma-sapinvoices-ui/blob/frontend-dev/webapp/utils/aws/ecs.py#L47-L53).

2. If you select `Review Run` and try to execute another run shortly afterwards, you'll be directed to a page indicating any active tasks (I think there can only be one at any given time) with a "monitoring" url that returns to the status page of the active task. 
    image

For additional questions/suggestions for improvements re: above, please wait until the next PR with the flask-ey bits! 🤓 Just wanted to provide another way to review this work.

**Review by ehanson8 — APPROVED** [review]

File: webapp/utils/aws/cloudwatch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ghukill) [reply webapp/utils/aws/cloudwatch.py]: Thanks for the reminder @jonavellecuerdo!  The missing link for me was forgetting that we use the `Config` object to hydrate the flask `app.config` here: https://github.com/MITLibraries/alma-sapinvoices-ui/blob/main/webapp/app.py#L8.

And, I'm assuming that only works when it's a _pre-defined_ property on the object, as there is no way for it to know which env vars _might_ be dynamically retrieved.
